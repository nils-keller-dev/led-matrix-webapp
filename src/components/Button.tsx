type ButtonProps = {
  text: string
  isSecondary?: boolean
  onClick: () => void
}

export function Button({ text, isSecondary, onClick }: ButtonProps) {
  return (
    <button
      className={`outline-none border rounded-md  px-4 py-2 ${isSecondary ? 'border-muted-foreground' : 'bg-primary text-background'}`}
      onClick={onClick}
    >
      {text}
    </button>
  )
}
