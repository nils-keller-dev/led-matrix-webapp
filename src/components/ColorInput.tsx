import { useSignal } from '@preact/signals'
import { useEffect, useRef } from 'preact/hooks'

type ColorInputProps = {
  initialValue: string
  onChange?: (color: string) => void
}

export function ColorInput({ initialValue, onChange }: ColorInputProps) {
  const colorInput = useRef<HTMLInputElement>(null)

  const currentColor = useSignal(initialValue)

  const onColorChange = () => {
    const color = colorInput.current?.value ?? ''
    if (!color || color === currentColor.value) return

    currentColor.value = color
    onChange?.(currentColor.value)
  }

  useEffect(() => {
    colorInput.current?.addEventListener('change', onColorChange)
  }, [])

  return (
    <div className="w-full h-10 rounded-md border border-secondary bg-background flex items-center relative">
      <label htmlFor="picker" className="size-full absolute" />
      <div
        className="size-5 rounded-md border border-secondary ml-4"
        style={{ backgroundColor: currentColor.value }}
      />
      <input
        id="picker"
        ref={colorInput}
        type="color"
        value={currentColor.value}
        className="opacity-0 w-0"
      />
      <span className="ml-3 uppercase text-primary">{currentColor.value}</span>
    </div>
  )
}
