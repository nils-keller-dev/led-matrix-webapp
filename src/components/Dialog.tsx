import * as RadixDialog from '@radix-ui/react-dialog'
import { JSX } from 'preact/jsx-runtime'

type DialogProps = {
  children: JSX.Element | false
  open: boolean
  onOpenChange: (open: boolean) => void
}

export function Dialog({ children, open, onOpenChange }: DialogProps) {
  return (
    <RadixDialog.Root open={open} onOpenChange={onOpenChange}>
      <RadixDialog.Portal>
        <RadixDialog.Overlay className="backdrop-brightness-50 backdrop-blur-xs size-screen fixed top-0 z-30" />
        <RadixDialog.Content
          onPointerDownOutside={(e) => e.preventDefault()}
          className="fixed top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 w-10/12 bg-background p-4 rounded-xl outline-hidden z-40 border border-secondary"
        >
          <RadixDialog.Title className="hidden" />
          {children}
        </RadixDialog.Content>
      </RadixDialog.Portal>
    </RadixDialog.Root>
  )
}
