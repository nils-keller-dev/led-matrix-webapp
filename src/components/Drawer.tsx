import { useSignal } from '@preact/signals'
import { JSX } from 'preact'
import { useEffect, useRef } from 'preact/hooks'

type DrawerProps = {
  children: JSX.Element | false
  header: JSX.Element
  isExpanded: boolean
  childrenId: number
  onChangeIsExpanded: (isExpanded: boolean) => void
}

const PADDING_BOTTOM = 30
const HEIGHT_PULLER = 52
const BACKDROP_BLUR = 3
const BACKDROP_BRIGHTNESS = 0.5

const getYPosition = (e: MouseEvent | TouchEvent) => {
  return e instanceof MouseEvent ? e.clientY : e.touches[0].clientY
}

const clamp = (value: number, min: number, max: number) => {
  return Math.max(min, Math.min(max, value))
}

export function Drawer({
  children,
  header,
  isExpanded,
  childrenId,
  onChangeIsExpanded
}: DrawerProps) {
  const drawerRef = useRef<HTMLDivElement>(null)
  const headerRef = useRef<HTMLDivElement>(null)
  const childrenRef = useRef<HTMLDivElement>(null)

  const expandedHeight = useSignal(0)
  const collapsedHeight = useSignal(0)
  const currentHeight = useSignal(0)
  const collapsedExpandedDelta = useSignal(0)

  const localIsExpanded = useSignal(isExpanded)

  const previousDragY = useSignal(0)

  useEffect(() => {
    if (isExpanded === localIsExpanded.value) return
    isExpanded ? expand() : collapse()
  }, [isExpanded])

  useEffect(() => {
    if (!headerRef.current || !childrenRef.current) return

    const headerHeight = headerRef.current.clientHeight + HEIGHT_PULLER
    const childrenHeight =
      childrenRef.current.clientHeight + (children ? PADDING_BOTTOM : 0)

    collapsedHeight.value = headerHeight
    expandedHeight.value = childrenHeight + headerHeight
    collapsedExpandedDelta.value = expandedHeight.value - collapsedHeight.value

    updateStyles(collapsedHeight.value)
  }, [childrenId])

  const updateStyles = (height: number) => {
    document.body.style.setProperty('--drawer-visible-height', `${height}px`)

    const percentOfExpanded = clamp(
      (height - collapsedHeight.value) / collapsedExpandedDelta.value,
      0,
      1
    )

    document.body.style.setProperty(
      '--drawer-backdrop-blur',
      `${percentOfExpanded * BACKDROP_BLUR}px`
    )

    document.body.style.setProperty(
      '--drawer-backdrop-brightness',
      (1 - percentOfExpanded * BACKDROP_BRIGHTNESS).toString()
    )

    currentHeight.value = height
  }

  const handleDrag = (e: MouseEvent | TouchEvent) => {
    const clientY = getYPosition(e)

    const calculatedHeight =
      currentHeight.value + (previousDragY.value - clientY)

    const newHeight = clamp(
      calculatedHeight,
      collapsedHeight.value - PADDING_BOTTOM,
      expandedHeight.value + PADDING_BOTTOM
    )

    updateStyles(newHeight)

    previousDragY.value = clientY
  }

  const startDrag = (e: MouseEvent | TouchEvent) => {
    previousDragY.value = getYPosition(e)
    drawerRef.current?.style.setProperty('transition', 'none')

    handleDrag(e)

    addEventListener('mousemove', handleDrag)
    addEventListener('touchmove', handleDrag)
    addEventListener('mouseup', stopDrag)
    addEventListener('touchend', stopDrag)
  }

  const stopDrag = () => {
    drawerRef.current?.style.removeProperty('transition')

    const distanceToExpanded = Math.abs(
      currentHeight.value - expandedHeight.value
    )

    const distanceToCollapsed = Math.abs(
      currentHeight.value - collapsedHeight.value
    )

    if (distanceToExpanded && distanceToCollapsed) {
      distanceToExpanded < distanceToCollapsed ? expand() : collapse()
    }

    removeEventListener('mousemove', handleDrag)
    removeEventListener('touchmove', handleDrag)
    removeEventListener('mouseup', stopDrag)
    removeEventListener('touchend', stopDrag)
  }

  const collapse = (e?: Event) => {
    e?.preventDefault()
    updateStyles(collapsedHeight.value)
    onChangeIsExpanded(false)
    localIsExpanded.value = false
  }

  const expand = () => {
    updateStyles(expandedHeight.value)
    onChangeIsExpanded(true)
    localIsExpanded.value = true
  }

  const toggle = () => {
    isExpanded ? collapse() : expand()
  }

  return (
    <div className="size-full absolute pointer-events-none">
      {currentHeight.value > collapsedHeight.value && (
        <div
          onTouchEnd={collapse}
          className="size-full backdrop-blur-[--drawer-backdrop-blur] backdrop-brightness-[--drawer-backdrop-brightness] pointer-events-auto"
        />
      )}
      <div
        ref={drawerRef}
        onTouchStart={startDrag}
        className="pb-[60px] w-full rounded-t-xl border border-secondary border-b-0 fixed top-full bg-background -translate-y-[--drawer-visible-height] transition-transform pointer-events-auto"
      >
        <div className="pt-4 mb-7" onClick={toggle}>
          <div className="bg-secondary w-24 h-2 rounded-full mx-auto" />
        </div>
        <div className="px-7">
          <div className="pb-7" ref={headerRef}>
            {header}
          </div>
          <div ref={childrenRef}>{children}</div>
        </div>
      </div>
    </div>
  )
}
