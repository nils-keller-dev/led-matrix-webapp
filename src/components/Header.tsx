export function Header() {
  return (
    <header className="border-b border-b-secondary flex items-center justify-between pr-4">
      <div className="flex items-center text-muted-foreground gap-1">
        <img src="/logo.svg" alt="logo" className="size-14" />
        <span>led-matrix-webapp</span>
      </div>
      <a href="/data" className="text-secondary">
        v1.0.0
      </a>
    </header>
  )
}
