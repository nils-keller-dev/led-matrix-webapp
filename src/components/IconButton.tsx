import { JSX, h } from "preact";

export function IconButton(props: JSX.HTMLAttributes<HTMLButtonElement>) {
  const styles =
    "size-9 rounded-full border border-secondary bg-background text-muted-foreground flex items-center justify-center transition-colors active:bg-secondary disabled:pointer-events-none disabled:opacity-50";
  const { children, ...buttonProps } = props;

  return (
    <button
      {...buttonProps}
      class={`${styles} ${buttonProps.class}`}
    >
      {children}
    </button>
  );
}
