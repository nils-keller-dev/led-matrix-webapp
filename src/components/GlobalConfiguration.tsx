import { Sun, SunDim } from 'lucide-preact'
import { Slider } from './Slider'
import { ColorInput } from './ColorInput'
import { rgbToHex } from '../utils/ColorConversion'

type GlobalConfigurationProps = {
  brightness: number
  color: number[]
  updateBrightness: (brightness: number) => void
  updateColor: (color: string) => void
}

export function GlobalConfiguration({
  brightness,
  color,
  updateBrightness,
  updateColor
}: GlobalConfigurationProps) {
  return (
    <div className="flex flex-col gap-5 w-full p-7">
      <div className="flex gap-3 flex-row">
        <SunDim class="text-muted-foreground shrink-0" />
        <Slider
          min={3}
          max={100}
          initialValue={brightness}
          onChange={updateBrightness}
        />
        <Sun class="text-muted-foreground shrink-0" />
      </div>
      <ColorInput initialValue={rgbToHex(color)} onChange={updateColor} />
    </div>
  )
}
