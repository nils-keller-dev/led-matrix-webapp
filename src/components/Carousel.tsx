import { useSignal } from "@preact/signals";
import { ArrowLeft, ArrowRight, Settings2 } from "lucide-preact";
import { IconButton } from "../components/IconButton.tsx";
import { h } from "preact";

interface CarouselItem {
  title: string;
  hasSettingsIcon: boolean;
}

interface CarouselProps {
  items: CarouselItem[];
  onClickSettings?: () => void;
}

export function Carousel({
  items,
  onClickSettings,
}: CarouselProps) {
  const activeItem = useSignal(0);

  const nextItem = () => {
    activeItem.value = (activeItem.value + 1) % items.length;
  };

  const previousItem = () => {
    activeItem.value = (activeItem.value - 1 + items.length) % items.length;
  };

  return (
    <div className="w-full p-7 aspect-square">
      <div className="text-7xl font-abril size-full  border border-secondary rounded-3xl flex items-center justify-center relative">
        {items[activeItem.value].title}
        <IconButton
          className="absolute top-1/2 left-0 -translate-x-1/2 -translate-y-1/2"
          onClick={previousItem}
        >
          <ArrowLeft />
        </IconButton>
        <IconButton
          className="absolute top-1/2 right-0 translate-x-1/2 -translate-y-1/2"
          onClick={nextItem}
        >
          <ArrowRight />
        </IconButton>
        {items[activeItem.value].hasSettingsIcon && (
          <IconButton
            className="absolute top-2 right-2"
            onClick={onClickSettings}
          >
            <Settings2 />
          </IconButton>
        )}
      </div>
      <div className="flex justify-center gap-1 mt-2">
        {items.map((_, index) => (
          <div
            class={`size-2 rounded-full ${
              index === activeItem.value
                ? "bg-muted-foreground"
                : "bg-secondary"
            }`}
          />
        ))}
      </div>
    </div>
  );
}
