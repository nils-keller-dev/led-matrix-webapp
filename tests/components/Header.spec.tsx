import { mount, shallow } from 'enzyme'
import { describe, expect, test, vi } from 'vitest'
import { Header } from '../../src/components/Header'

vi.mock('../../package.json', () => ({
  default: { version: '0.0.0' }
}))

describe('Header', () => {
  test('matches snapshot', () => {
    const wrapper = mount(<Header />)
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('renders with version number', () => {
    const wrapper = shallow(<Header />)

    expect(wrapper.text()).toContain('v0.0.0')
  })
})
