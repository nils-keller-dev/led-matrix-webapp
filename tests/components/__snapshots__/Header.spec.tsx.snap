// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > matches snapshot 1`] = `"<header class="border-b border-b-secondary flex items-center justify-between pr-4"><div class="flex items-center text-muted-foreground gap-1"><img src="/logo.svg" alt="logo" class="size-14"><span>led-matrix-webapp</span></div><a href="/api/state" class="text-secondary">v0.0.0</a></header>"`;
